topics/functions-intro.md

```diff
@@ -0,0 +1,353 @@
+
+By design, Redis only caches the loaded scripts.
+That means that the script cache can become lost at any time, such as after calling `SCRIPT FLUSH` or a server restart.
+The application is responsible for reloading scripts during runtime if any are missing.
+The underlying assumption is that scripts are a part of the application and not maintained by the Redis server.
+
+This approach suits many light-weight scripting use cases but introduces several difficulties once an application becomes complex and relies more heavily on scripting, namely:
+
+1. All client application instances must maintain a copy of all scripts. That means having some mechanism that applies script updates to all of the application's instances.
+1. Calling cached scripts within the context of a [transaction](/topics/transactions) increases the probability of the transaction failing because of a missing script. Being more likely to fail makes using cached scripts as building blocks of workflows less attractive.
+1. SHA1 digests are meaningless, making debugging the system extremely hard (e.g., in a `MONITOR` session).
+1. When used naively, `EVAL` promotes an anti-pattern in which scripts the client application renders verbatim scripts instead of responsibly using the [`!KEYS` and `ARGV` Lua APIs](/topics/lua-api#runtime-globals).
+1. Because they are ephemeral, a script can't call another script. This makes sharing and reusing code between scripts nearly impossible, short of client-side preprocessing (see the first point).
+
+To address these needs while avoiding breaking changes to already-established and well-liked ephemeral scripts, Redis v7.0 introduces Redis Functions.
+
+## What are Redis Functions?
+
+Redis functions are an evolutionary step from ephemeral scripting.
+
+Functions provide the same core functionality as scripts but are first-class software artifacts citizens of the database.
+Redis manages functions as an integral part of the database and ensures their availability via data persistence and replication.
+Because functions are part of the database and therefore declared before use, applications aren't required to load them during runtime nor risk aborted transactions.
+An application that uses functions depends only on their APIs rather than on the embedded script logic in the database.
+
+Whereas ephemeral scripts are considered a part of the application's domain, functions extend the database server itself with user-provided logic.
+Every function has a unique user-provided name, making it much easier to call and trace its execution.
+
+The design of Redis Functions also attempts to demarcate between the programming language used for writing functions and their management by the server.
```

Comment:
i'm still missing something in this intro that stresses that functions are here as an alternative to a schema, so if you have several keys in which the data should always be consistent, and they're always updated together, the application is expected expected to use functions to update the data, and the functions will present logical API for the data (which means the application is not expected to issue any redis native commands to access it, only the high level logical function calls).

and maybe some paragraph that compares the to modules, developed once, loaded at startup, and used repeatedly by various applications / clients.

Reply:
It is explained on the function-into page.

Reply:
this is the functions-into page.

anyway, i think the key differences between the two scripts mechanisms, and what's each one should be used for, needs to be specified in both the programability into page, and each of the specific pages (to make users re-consider going to the other one).
maybe the level of detail should be different (less detail in the programability page, and more details in the functions page)